### ISSUE TITLE
AI: hyperparameter, metric, metricDecisionThreshold - can be multiple. Property talks about only one.

### ISSUE THREAD

**Issue by VenkatTechnologist** [spdx/spdx-3-model#769]
There could be multiple hyperparameters, metrics, and metricDecisionThresholds for an AI model.
The properties 'hyperparameter', 'metric' and 'metricDecisionThreshold' talk about only one value, and their class
is DictionaryEntry, which can hold only one key-pair. How can we accommodate multiple hyperparameters,
metrics, and metricDecisionThresholds of an AI model in the AI profile of SPDX?

**Comment by bennetkl** [comment]
We decided for first release of AI BOM  to keep it relatively simple, but true we should consider this functionality in 3.1.  Do you have an use case /  example of multiple  hyperparameters in use (ideally an open source project). 

    On Thursday, May 9, 2024 at 07:13:23 a.m. EDT, VenkatTechnologist ***@***.***> wrote:  
 
 


There could be multiple hyperparameters while training an AI model.
The property 'hyperparameter' talks about one value, and it's data type
is dictionary, which has minValue as 1 and maxValue as 1. How can we
accommodate multiple hyperparameters of an AI model in the AI profile
of SPDX?

**Comment by VenkatTechnologist** [comment]
Some hyperparameter examples:

Machine Learning:

- No. of clusters
- Distance
- Minimum points required to be considered a cluster
- Initialization method
- L1/L2 regularization

Generative-AI:

- No. of layers in the neural network
- No. of units (neurons) per layer
- Learning rate
- Loss function
- L1/L2 regularization
- Neuron dropout
- Number of training iterations
- Data sample batch size
- Latent Space Dimension in VAE
- Inception score in Image generation

The implementation can pick and choose multiple hyperparameters for implementation.

**Comment by VenkatTechnologist** [comment]
Metrics used to assess and their associated thresholds can be multiple in machine learning as well as generative AI algorithms.

Machine learning:

- Classification
   - Accuracy
   - Precision
   - Recall
   - F1-Score
   - AUC-RAC
   - Log Loss

- Regression
   - MSE
   - RMSE
   - MAE
   - R-squared

Generative AI:

- Inception score
- Fréchet Inception Distance

**Comment by VenkatTechnologist** [comment]
On that note, it looks like we need a class called 'Dictionary' to list multiple hyperparameters, metrics, and metric thresholds.

**Comment by bact** [comment]
In [`AIPackage`](https://github.com/spdx/spdx-3-model/blob/main/model/AI/Classes/AIPackage.md), all of them (hyperparameter, metric, metricDecisionThreshold) has `minCount` = 0 and no `maxCount`. Which means one `AIPackage` can have multiple of these properties (= multiple entries).

**Comment by bact** [comment]
I think we can close this one, as it is very clear that the current model can accommodate the expressed concern.

**Comment by VenkatTechnologist** [comment]
Ok

**Comment by VenkatTechnologist** [comment]
Having said that, I wonder how it would work with systems with multiple AI
components.
How would we map which metric, threshold, and hyperparameters to individual
AI component?

**Comment by bact** [comment]
It will be through relationships like `contains` and `dependsOn`.

See possible relationship types here: https://spdx.github.io/spdx-spec/v3.0/model/Core/Vocabularies/RelationshipType/

**Comment by bact** [comment]
@goneall I think we can close this one as it is a non-issue: the cardinality of this property in AIPackage is 0..* -- @VenkatTechnologist has agreed on this.

**Comment by goneall** [comment]
Closing per above suggestion
